File: persistence/persistence_client/persistence_client.go
package persistence_client

import (
	"encoding/json"
	"io"

	"strconv"

	"context"

	"github.com/davecourtois/Globular/api"
	"github.com/davecourtois/Globular/persistence/persistencepb"
	"github.com/davecourtois/Utility"
	"google.golang.org/grpc"
)

////////////////////////////////////////////////////////////////////////////////
// Persitence Client Service
////////////////////////////////////////////////////////////////////////////////
type Persistence_Client struct {
	cc *grpc.ClientConn
	c  persistencepb.PersistenceServiceClient

	// The id of the service
	id string

	// The name of the service
	name string

	// The client domain
	domain string

	// The port
	port int

	// is the connection is secure?
	hasTLS bool

	// Link to client key file
	keyFile string

	// Link to client certificate file.
	certFile string

	// certificate authority file
	caFile string
}

// Create a connection to the service.
func NewPersistence_Client(address string, id string) (*Persistence_Client, error) {
	client := new(Persistence_Client)
	err := api.InitClient(client, address, id)
	if err != nil {
		return nil, err
	}
	client.cc, err = api.GetClientConnection(client)
	if err != nil {
		return nil, err
	}
	client.c = persistencepb.NewPersistenceServiceClient(client.cc)

	return client, nil
}

func (self *Persistence_Client) Invoke(method string, rqst interface{}, ctx context.Context) (interface{}, error) {
	if ctx == nil {
		ctx = api.GetClientContext(self)
	}
	return api.InvokeClientRequest(self.c, ctx, method, rqst)
}

// Return the domain
func (self *Persistence_Client) GetDomain() string {
	return self.domain
}

// Return the address
func (self *Persistence_Client) GetAddress() string {
	return self.domain + ":" + strconv.Itoa(self.port)
}

// Return the id of the service instance
func (self *Persistence_Client) GetId() string {
	return self.id
}

// Return the name of the service
func (self *Persistence_Client) GetName() string {
	return self.name
}

// must be close when no more needed.
func (self *Persistence_Client) Close() {
	if self.cc != nil {
		self.cc.Close()
	}
}

// Set grpc_service port.
func (self *Persistence_Client) SetPort(port int) {
	self.port = port
}

// Set the client instance id.
func (self *Persistence_Client) SetId(id string) {
	self.id = id
}

// Set the client name.
func (self *Persistence_Client) SetName(name string) {
	self.name = name
}

// Set the domain.
func (self *Persistence_Client) SetDomain(domain string) {
	self.domain = domain
}

////////////////// TLS ///////////////////

// Get if the client is secure.
func (self *Persistence_Client) HasTLS() bool {
	return self.hasTLS
}

// Get the TLS certificate file path
func (self *Persistence_Client) GetCertFile() string {
	return self.certFile
}

// Get the TLS key file path
func (self *Persistence_Client) GetKeyFile() string {
	return self.keyFile
}

// Get the TLS key file path
func (self *Persistence_Client) GetCaFile() string {
	return self.caFile
}

// Set the client is a secure client.
func (self *Persistence_Client) SetTLS(hasTls bool) {
	self.hasTLS = hasTls
}

// Set TLS certificate file path
func (self *Persistence_Client) SetCertFile(certFile string) {
	self.certFile = certFile
}

// Set TLS key file path
func (self *Persistence_Client) SetKeyFile(keyFile string) {
	self.keyFile = keyFile
}

// Set TLS authority trust certificate file path
func (self *Persistence_Client) SetCaFile(caFile string) {
	self.caFile = caFile
}

///////////////////////// API /////////////////////

// Create a new datastore connection.
func (self *Persistence_Client) CreateConnection(connectionId string, name string, host string, port float64, storeType float64, user string, pwd string, timeout float64, options string, save bool) error {
	rqst := &persistencepb.CreateConnectionRqst{
		Connection: &persistencepb.Connection{
			Id:       connectionId,
			Name:     name,
			Host:     host,
			Port:     int32(Utility.ToInt(port)),
			Store:    persistencepb.StoreType(storeType),
			User:     user,
			Password: pwd,
			Timeout:  int32(Utility.ToInt(timeout)),
			Options:  options,
		},
		Save: save,
	}

	_, err := self.c.CreateConnection(api.GetClientContext(self), rqst)
	return err
}

func (self *Persistence_Client) DeleteConnection(connectionId string) error {
	rqst := &persistencepb.DeleteConnectionRqst{
		Id: connectionId,
	}

	_, err := self.c.DeleteConnection(api.GetClientContext(self), rqst)
	return err
}

func (self *Persistence_Client) Connect(id string, password string) error {
	rqst := &persistencepb.ConnectRqst{
		ConnectionId: id,
		Password:     password,
	}

	_, err := self.c.Connect(api.GetClientContext(self), rqst)
	return err
}

func (self *Persistence_Client) Disconnect(connectionId string) error {

	rqst := &persistencepb.DisconnectRqst{
		ConnectionId: connectionId,
	}

	_, err := self.c.Disconnect(api.GetClientContext(self), rqst)

	return err
}

func (self *Persistence_Client) Ping(connectionId string) error {

	rqst := &persistencepb.PingConnectionRqst{
		Id: connectionId,
	}

	_, err := self.c.Ping(api.GetClientContext(self), rqst)

	return err
}

func (self *Persistence_Client) FindOne(connectionId string, database string, collection string, jsonStr string, options string) (string, error) {

	// Retreive a single value...
	rqst := &persistencepb.FindOneRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		Query:      jsonStr,
		Options:    options,
	}

	rsp, err := self.c.FindOne(api.GetClientContext(self), rqst)

	if err != nil {
		return "", err
	}

	return rsp.GetJsonStr(), err
}

func (self *Persistence_Client) Find(connectionId string, database string, collection string, query string, options string) (string, error) {

	// Retreive a single value...
	rqst := &persistencepb.FindRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		Query:      query,
		Options:    options,
	}

	stream, err := self.c.Find(api.GetClientContext(self), rqst)

	if err != nil {
		return "", err
	}

	valuesStr := "["
	for {
		results, err := stream.Recv()
		if err != nil {
			if err == io.EOF {
				// end of stream...
				break
			} else {
				return "", err
			}
		} else {
			if len(valuesStr) > 1 {
				valuesStr += ","
			}
			valuesStr += results.JsonStr[1 : len(results.JsonStr)-1]
		}

	}

	valuesStr += "]"

	return valuesStr, nil
}

/**
 * Usefull function to query and transform document.
 */
func (self *Persistence_Client) Aggregate(connectionId, database string, collection string, pipeline string, options string) (string, error) {
	// Retreive a single value...
	rqst := &persistencepb.AggregateRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		Pipeline:   pipeline,
		Options:    options,
	}

	stream, err := self.c.Aggregate(api.GetClientContext(self), rqst)

	if err != nil {
		return "", err
	}

	valuesStr := "["
	for {
		results, err := stream.Recv()
		if err != nil {
			if err == io.EOF {
				// end of stream...
				break
			} else {
				return "", err
			}
		} else {
			if len(valuesStr) > 1 {
				valuesStr += ","
			}
			valuesStr += results.JsonStr[1 : len(results.JsonStr)-1]
		}

	}

	valuesStr += "]"

	return string(valuesStr), nil
}

/**
 * Count the number of document that match the query.
 */
func (self *Persistence_Client) Count(connectionId string, database string, collection string, query string, options string) (int, error) {

	rqst := &persistencepb.CountRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		Query:      query,
		Options:    options,
	}

	rsp, err := self.c.Count(api.GetClientContext(self), rqst)

	if err != nil {
		return 0, err
	}

	return int(rsp.Result), err
}

/**
 * Insert one value in the database.
 */
func (self *Persistence_Client) InsertOne(connectionId string, database string, collection string, jsonStr string, options string) (string, error) {

	rqst := &persistencepb.InsertOneRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		JsonStr:    jsonStr,
		Options:    options,
	}

	rsp, err := self.c.InsertOne(api.GetClientContext(self), rqst)

	if err != nil {
		return "", err
	}

	return rsp.GetId(), err
}

func (self *Persistence_Client) InsertMany(connectionId string, database string, collection string, jsonStr string, options string) (string, error) {

	stream, err := self.c.InsertMany(api.GetClientContext(self))
	if err != nil {
		return "", err
	}

	// here you must run the sql service test before runing this test in order
	// to generate the file Employees.json
	data := make([]map[string]interface{}, 0)

	err = json.Unmarshal([]byte(jsonStr), &data)
	if err != nil {
		return "", err
	}

	// Persist 500 rows at time to save marshaling unmarshaling cycle time.
	chunkSize := 500

	for i := 0; i < len(data); i += chunkSize {

		var jsonStr []byte
		if i+chunkSize < len(data) {
			jsonStr, err = json.Marshal(data[i : i+chunkSize])
			if err != nil {
				return "", err
			}
		} else {
			jsonStr, err = json.Marshal(data[i:])
			if err != nil {
				return "", err
			}
		}

		rqst := &persistencepb.InsertManyRqst{
			Id:         connectionId,
			Database:   database,
			Collection: collection,
			JsonStr:    "[" + string(jsonStr) + "]",
		}

		err = stream.Send(rqst)
		if err != nil {
			return "", err
		}
	}

	rsp, err := stream.CloseAndRecv()
	if err != nil {
		return "", err
	}

	return rsp.Ids, nil
}

/**
 * Insert one value in the database.
 */
func (self *Persistence_Client) ReplaceOne(connectionId string, database string, collection string, query string, value string, options string) error {

	rqst := &persistencepb.ReplaceOneRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		Query:      query,
		Value:      value,
		Options:    options,
	}

	_, err := self.c.ReplaceOne(api.GetClientContext(self), rqst)

	return err
}

func (self *Persistence_Client) UpdateOne(connectionId string, database string, collection string, query string, value string, options string) error {

	rqst := &persistencepb.UpdateOneRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		Query:      query,
		Value:      value,
		Options:    options,
	}

	_, err := self.c.UpdateOne(api.GetClientContext(self), rqst)

	return err
}

/**
 * Update one or more document.
 */
func (self *Persistence_Client) Update(connectionId string, database string, collection string, query string, value string, options string) error {

	rqst := &persistencepb.UpdateRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		Query:      query,
		Value:      value,
		Options:    options,
	}

	_, err := self.c.Update(api.GetClientContext(self), rqst)

	return err
}

/**
 * Delete one document from the db
 */
func (self *Persistence_Client) DeleteOne(connectionId string, database string, collection string, query string, options string) error {

	rqst := &persistencepb.DeleteOneRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		Query:      query,
		Options:    options,
	}

	_, err := self.c.DeleteOne(api.GetClientContext(self), rqst)

	if err != nil {
		return err
	}

	return err
}

/**
 * Delete many document from the db.
 */
func (self *Persistence_Client) Delete(connectionId string, database string, collection string, query string, options string) error {

	rqst := &persistencepb.DeleteRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
		Query:      query,
		Options:    options,
	}

	_, err := self.c.Delete(api.GetClientContext(self), rqst)

	if err != nil {
		return err
	}

	return err
}

/**
 * Drop a collection.
 */
func (self *Persistence_Client) DeleteCollection(connectionId string, database string, collection string) error {
	// Test drop collection.
	rqst_drop_collection := &persistencepb.DeleteCollectionRqst{
		Id:         connectionId,
		Database:   database,
		Collection: collection,
	}
	_, err := self.c.DeleteCollection(api.GetClientContext(self), rqst_drop_collection)

	return err
}

/**
 * Drop a database.
 */
func (self *Persistence_Client) DeleteDatabase(connectionId string, database string) error {
	// Test drop collection.
	rqst_drop_db := &persistencepb.DeleteDatabaseRqst{
		Id:       connectionId,
		Database: database,
	}

	_, err := self.c.DeleteDatabase(api.GetClientContext(self), rqst_drop_db)

	return err
}

/**
 * Admin function, that must be protected.
 */
func (self *Persistence_Client) RunAdminCmd(connectionId string, user string, pwd string, script string) error {
	// Test drop collection.
	rqst_drop_db := &persistencepb.RunAdminCmdRqst{
		ConnectionId: connectionId,
		Script:       script,
		User:         user,
		Password:     pwd,
	}

	_, err := self.c.RunAdminCmd(api.GetClientContext(self), rqst_drop_db)

	return err
}

File: persistence/persistence_client/csharp/GlobularPersistenceClient/persistence_client.cs
using Grpc.Core;
using System;
using System.Threading.Tasks;
using System.Collections;
using System.Text.Json;

namespace Globular
{
    public class PersistenceClient : Client
    {
        private Persistence.PersistenceService.PersistenceServiceClient client;

        /// <summary>
        /// gRPC client for persistence service.
        /// </summary>
        /// <param name="address">Can be a domain or a IP address ex: localhos or 127.0.0.1</param>
        /// <param name="name">The name of the service on the server. ex: persistence_server</param>
        /// <returns>Return the instance of the client with it connection ready to be use.</returns>
        public PersistenceClient(string address, string name) : base(address, name)
        {
            // Here I will create grpc connection with the service...
            this.client = new Persistence.PersistenceService.PersistenceServiceClient(this.channel);
        }

        /// <summary>
        /// Create a new persistence connection
        /// </summary>
        /// <param name="connection">The connection information</param>
        /// <param name="save">If true the connection will be save in the configuation file.</param>
        public void CreateConnection(Persistence.Connection connection, bool save)
        {
            // Here I will create the new connection.
            Persistence.CreateConnectionRqst rqst = new Persistence.CreateConnectionRqst();
            rqst.Connection = connection;
            rqst.Save = save;

            // Create a new connection
            this.client.CreateConnection(rqst, this.GetClientContext());
        }

        /// <summary>
        /// Delete a connection with a given id.
        /// </summary>
        /// <param name="connectionId">The connection to delete</param>
        public void DeleteConnection(string connectionId)
        {
            var rqst = new Persistence.DeleteConnectionRqst();
            rqst.Id = connectionId;
            this.client.DeleteConnection(rqst, this.GetClientContext());
        }

        /// <summary>
        /// Open a connection with the datastore.
        /// </summary>
        /// <param name="connectionId"></param>
        public void Connect(string connectionId)
        {
            var rqst = new Persistence.ConnectRqst();
            rqst.ConnectionId = connectionId;
            this.client.Connect(rqst, this.GetClientContext());
        }

        /// <summary>
        /// Disconnect from the  server.
        /// </summary>
        /// <param name="connectionId">The connection id</param>
        public void Disconnect(string connectionId)
        {
            var rqst = new Persistence.DisconnectRqst();
            rqst.ConnectionId = connectionId;
            this.client.Disconnect(rqst, this.GetClientContext());
        }

        /// <summary>
        /// Ping a given persistence service.
        /// </summary>
        /// <param name="connectionId">The connection id, not it name.</param>
        /// <returns>Must return 'pong'</returns>
        public string Ping(string connectionId)
        {
            // Here I will create the new connection.
            Persistence.PingConnectionRqst rqst = new Persistence.PingConnectionRqst();
            rqst.Id = connectionId;

            // Create a new connection
            var rsp = this.client.Ping(rqst, this.GetClientContext());
            return rsp.Result;
        }

        ///////////////////////////////////// Quering /////////////////////////////////////

        /// <summary>
        /// Find one object from the database.
        /// </summary>
        /// <param name="connectionId">The connection id</param>
        /// <param name="database">The database name</param>
        /// <param name="collection">The collection name</param>
        /// <param name="query">The filter</param>
        /// <param name="options">a list of option, must be a json array</param>
        /// <returns></returns>
        public string FindOne(string connectionId, string database, string collection, string query, string options)
        {
            var rqst = new Persistence.FindOneRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.Query = query;
            rqst.Options = options;

            var rsp = this.client.FindOne(rqst, this.GetClientContext());
            return rsp.JsonStr;
        }

        /// <summary>
        /// Find multiple values from the data store.
        /// </summary>
        /// <param name="connectionId">The connection Id to be used</param>
        /// <param name="database">The database name</param>
        /// <param name="collection">The collection name</param>
        /// <param name="query">The query</param>
        /// <param name="options">a list of option, must be a json array</param>
        /// <returns></returns>
        public string Find(string connectionId, string database, string collection, string query, string options)
        {
            var rqst = new Persistence.FindRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.Query = query;
            rqst.Options = options;

            var call = this.client.Find(rqst, this.GetClientContext());

            // Make the function synchrone...
            string jsonStr = "[";
            bool hasNext = true;

            // read until no more values found...
            while (hasNext)
            {
                var task = Task.Run(() => call.ResponseStream.MoveNext());
                task.Wait(); // wait for the next value...
                hasNext = task.Result;
                if (hasNext)
                {
                    string str = call.ResponseStream.Current.JsonStr;
                    if (jsonStr.Length > 1)
                    {
                        jsonStr += ",";
                    }
                    jsonStr += str.Substring(1, str.Length - 1);
                }
            }

            return jsonStr + "]";
        }

        public string Aggregate(string connectionId, string database, string collection, string pipeline, string options)
        {
            var rqst = new Persistence.AggregateRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.Pipeline = pipeline;
            rqst.Options = options;

            var call = this.client.Aggregate(rqst, this.GetClientContext());

            // Make the function synchrone...
            string jsonStr = "[";
            bool hasNext = true;

            // read until no more values found...
            while (hasNext)
            {
                var task = Task.Run(() => call.ResponseStream.MoveNext());
                task.Wait(); // wait for the next value...
                hasNext = task.Result;
                if (hasNext)
                {
                    string str = call.ResponseStream.Current.JsonStr;
                    if (jsonStr.Length > 1)
                    {
                        jsonStr += ",";
                    }
                    jsonStr += str.Substring(1, str.Length - 1);
                }
            }

            return jsonStr + "]";
        }

        /// <summary>
        /// Count the number of document that match a given query
        /// </summary>
        /// <param name="connectionId">The connection id</param>
        /// <param name="database">The datase</param>
        /// <param name="collection">The collection</param>
        /// <param name="query">The query</param>
        /// <param name="options">A list of options in form of json string</param>
        /// <returns></returns>
        public long Count(string connectionId, string database, string collection, string query, string options)
        {
            var rqst = new Persistence.CountRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.Query = query;
            rqst.Options = options;

            var rsp = this.client.Count(rqst, this.GetClientContext());
            return rsp.Result;
        }

        /// <summary>
        /// Insert one document in the database and return the newly create document id.
        /// </summary>
        /// <param name="connectionId">The connection id</param>
        /// <param name="database">The database name</param>
        /// <param name="collection">The collection</param>
        /// <param name="jsonStr">The oject stringnify value</param>
        /// <param name="options">The options</param>
        /// <returns></returns>
        public string InsertOne(string connectionId, string database, string collection, string jsonStr, string options)
        {
            var rqst = new Persistence.InsertOneRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.JsonStr = jsonStr;
            rqst.Options = options;

            var rsp = this.client.InsertOne(rqst, this.GetClientContext());
            return rsp.Id;
        }

        public void ReplaceOne(string connectionId, string database, string collection, string query, string value, string options)
        {
            var rqst = new Persistence.ReplaceOneRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.Query = query;
            rqst.Value = value;
            rqst.Options = options;

            this.client.ReplaceOne(rqst, this.GetClientContext());
        }

        public void UpdateOne(string connectionId, string database, string collection, string query, string value, string options)
        {
            var rqst = new Persistence.UpdateOneRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.Query = query;
            rqst.Value = value;
            rqst.Options = options;

            this.client.UpdateOne(rqst, this.GetClientContext());
        }


        public void Update(string connectionId, string database, string collection, string query, string value, string options)
        {
            var rqst = new Persistence.UpdateRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.Query = query;
            rqst.Value = value;
            rqst.Options = options;

            this.client.Update(rqst, this.GetClientContext());
        }

        public void DeleteOne(string connectionId, string database, string collection, string query, string options)
        {
            var rqst = new Persistence.DeleteOneRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.Query = query;
            rqst.Options = options;

            this.client.DeleteOne(rqst, this.GetClientContext());
        }

        public void Delete(string connectionId, string database, string collection, string query, string options)
        {
            var rqst = new Persistence.DeleteRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;
            rqst.Query = query;
            rqst.Options = options;

            this.client.Delete(rqst, this.GetClientContext());
        }

        public string InsertMany(string connectionId, string database, string collection, ArrayList objects, string jsonStr, string options)
        {

            // Open a stream with the server.
            var call = this.client.InsertMany(this.GetClientContext());

            // Here i will iterate over the list of object contain in the collection and persist 500 object at time.
            var chunkSize = 500;
            for (var i = 0; i < objects.Count; i += chunkSize)
            {
                var rqst = new Persistence.InsertManyRqst();
                rqst.Id = connectionId;
                rqst.Database = database;
                rqst.Collection = collection;

                if (i + chunkSize < objects.Count)
                {
                    rqst.JsonStr = "[" + JsonSerializer.Serialize(objects.GetRange(i, chunkSize)) + "]";
                }
                else
                {
                    rqst.JsonStr = "[" + JsonSerializer.Serialize(objects.GetRange(i, objects.Count - i)) + "]";
                }

                var task = Task.Run(() => call.RequestStream.WriteAsync(rqst));
                task.Wait(); // wait until the message was sent...
            }

            Task.Run(() => call.RequestStream.CompleteAsync()).Wait();

            var rsp = Task.Run(() => call.ResponseAsync);
            rsp.Wait();

            return rsp.Result.Ids;
        }

        public void DeleteCollection(string connectionId, string database, string collection)
        {
            var rqst = new Persistence.DeleteCollectionRqst();
            rqst.Id = connectionId;
            rqst.Database = database;
            rqst.Collection = collection;

            this.client.DeleteCollection(rqst, this.GetClientContext());
        }

        public void DeleteDatabase(string connectionId, string database)
        {
            var rqst = new Persistence.DeleteDatabaseRqst();
            rqst.Id = connectionId;
            rqst.Database = database;

            this.client.DeleteDatabase(rqst, this.GetClientContext());
        }

        public void RunAdminCmd(string connectionId, string user, string pwd, string script)
        {
            var rqst = new Persistence.RunAdminCmdRqst();
            rqst.ConnectionId = connectionId;
            rqst.Script = script;
            rqst.User = user;
            rqst.Password = pwd;

            this.client.RunAdminCmd(rqst, this.GetClientContext());
        }
        
    }

}
